src/app/reporting/ReportingEngine.cpp

```diff
@@ -0,0 +1,327 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements reporting engine for CHIP
+ *      Data Model profile.
+ *
+ */
+
+#include <app/InteractionModelEngine.h>
+#include <app/reporting/LoggingManagement.h>
+#include <app/reporting/ReportingEngine.h>
+
+using namespace ::chip;
+using namespace ::chip::TLV;
+
+namespace chip {
+namespace app {
+namespace reporting {
+CHIP_ERROR ReportingEngine::Init()
+{
+    CHIP_ERROR err       = CHIP_NO_ERROR;
+    mMoreChunkedMessages = false;
+    mNumReportsInFlight  = 0;
+    mCurReadHandlerIdx   = 0;
+    return err;
+}
+
+CHIP_ERROR ReportingEngine::BuildSingleReportDataEventList(ReportData::Builder & aReportDataBuilder, ReadHandler * apReadHandler)
+{
+    CHIP_ERROR err                = CHIP_NO_ERROR;
+    chip::EventNumber event_count = 0;
+    bool dataClean;
+    chip::EventNumber initialEvents[kPriorityLevel_Last - kPriorityLevel_First + 1];
+
+    memcpy(initialEvents, apReadHandler->mSelfVendedEvents, sizeof(initialEvents));
+
+    EventList::Builder eventList = aReportDataBuilder.CreateEventDataListBuilder();
+
+    // events only enter the picture if the read handler is
+    // subscribed to events.
+    // TODO:: Add mechanism to get particular set of interested clusters instead of all events
+    if (apReadHandler->mGetToAllEvents)
+    {
+        // Verify that we have events to transmit
+        LoggingManagement & logger = LoggingManagement::GetInstance();
+
+        // If the logger is not valid or has not been initialized,
+        // skip the rest of processing
+        VerifyOrExit(logger.IsValid(), );
+
+        for (int i = 0; i < kPriorityLevel_Last - kPriorityLevel_First + 1; i++)
```

Comment:
since we static cast `i` to `PriorityLevel` anyway, is theree a benefit for the comparison with a delta? We seem to implicityly `static_assert(kPriorityLevel_First == 0)`.

How about just using `i < kPriorityLevelLast + 1` for comparison?

Otherwise it seems like the loop should be fr om kPriorityLevelFirst to kPriorityLevelLast, but then we would need separate indexing into initialEvents.

Reply:
agree, done, thanks